server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferWithSkipList.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.uniffle.server.buffer;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentNavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Lists;
+import org.roaringbitmap.longlong.Roaring64NavigableMap;
+
+import org.apache.uniffle.common.BufferSegment;
+import org.apache.uniffle.common.ShuffleDataDistributionType;
+import org.apache.uniffle.common.ShufflePartitionedBlock;
+import org.apache.uniffle.common.ShufflePartitionedData;
+import org.apache.uniffle.common.util.Constants;
+import org.apache.uniffle.common.util.JavaUtils;
+import org.apache.uniffle.server.ShuffleDataFlushEvent;
+import org.apache.uniffle.server.ShuffleFlushManager;
+
+public class ShuffleBufferWithSkipList extends ShuffleBuffer {
+  private ConcurrentSkipListMap<Long, ShufflePartitionedBlock> blocksMap;
+  private final Map<Long, ConcurrentSkipListMap<Long, ShufflePartitionedBlock>> inFlushBlockMap;
+
+  public ShuffleBufferWithSkipList(long capacity) {
+    super(capacity);
+    this.blocksMap = newConcurrentSkipListMap();
+    this.inFlushBlockMap = JavaUtils.newConcurrentMap();
+  }
+
+  private ConcurrentSkipListMap<Long, ShufflePartitionedBlock> newConcurrentSkipListMap() {
+    // Sort by taskAttemptId.
```

Comment:
We put `blockId` into `blocksMap`, why it is `taskAttemptId` in the comment?

Reply:
We just need to ensure the order of taskAttemptId here for we need sort blocks when flush. And the block id is reversed when making comparisons.

Reply:
I don't get it. You sort ConcurrentSkipListMap by revered blockId. How is it related to taskAttemptId?

Reply:
taskAttemptId is in the lowest bits of block id.
https://github.com/apache/incubator-uniffle/blob/41d78aafac933e7c7ffa5dddd440b7465a3bb63e/common/src/main/java/org/apache/uniffle/common/util/BlockIdLayout.java#L28-L30

Reply:
Please note this in the comment too.